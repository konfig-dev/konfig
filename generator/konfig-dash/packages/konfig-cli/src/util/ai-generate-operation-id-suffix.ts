import { Operation, OperationObject } from 'konfig-lib'
import OpenAI from 'openai'

export async function aiGenerateOperationIdSuffix({
  prefix,
  path,
  method,
  operation,
  operations,
}: {
  prefix: string
  path: string
  method: string
  operation: Operation
  operations: OperationObject[]
}): Promise<string> {
  const openai = new OpenAI({ apiKey: process.env['OPENAI_API_KEY'] })

  const existingIds = operations.map((o) => o.operation.operationId)
  const existingIdsStr = existingIds.join(', ').slice(0, -2)
  const systemPrompt = `You are a generator of operation IDs for an OpenAPI specification. You generate operation IDs in json format: "{ prefix: ... , suffix: ...}"`
  const prompt = generatePrompt({
    prefix,
    path,
    method,
    operation,
    existingIdsStr,
  })

  const response = (
    await openai.chat.completions.create({
      response_format: { type: 'json_object' },
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt },
      ],
      model: 'gpt-3.5-turbo-1106',
    })
  ).choices[0].message.content

  if (response === undefined || response === null || response === '')
    throw Error('OpenAI returned an empty response')

  const responseJson = JSON.parse(response)
  if (responseJson === undefined || responseJson === null)
    throw Error('OpenAI returned an invalid response: it is not a valid json')
  if (responseJson.suffix === undefined || responseJson.suffix === null)
    throw Error('OpenAI returned an invalid json: it does not contain a suffix')

  // Post-process to ensure that rules are followed
  let suffix = responseJson.suffix.replace('_', '')
  suffix = suffix.charAt(0).toLowerCase() + suffix.slice(1)
  suffix = suffix.replace('retrieve', 'get')
  suffix = suffix.replace('fetch', 'get')
  console.log(`AI-generated operation id: ${prefix}${suffix}`)
  return suffix
}

function generatePrompt({
  prefix,
  path,
  method,
  operation,
  existingIdsStr,
}: {
  prefix: string
  path: string
  method: string
  operation: Operation
  existingIdsStr: string
}): string {
  return `Generate an operation ID for this operation in my OpenAPI specification:
Path: ${path}
Method: ${method}
Summary: ${operation.summary}
Description: ${operation.description}

The operation ID is in the format "prefix_suffix". The prefix is: ${prefix}. Generate the suffix and respond in json format.
The operation ID MUST follow ALL of these rules:

1. The suffix must be at least 3 characters long.
2. The suffix must be in camelCase and CANNOT include any underscores.
3. The suffix must not redundantly re-use words from the prefix. (For example: Store_get is preferred over Store_getStore.)
4. If the suffix contains both a verb and a noun, the noun should generally follow the verb. (For example, Store_placeOrder is better than Store_orderPlace).
5. Prefer simple verbs like "get", "create", "update", "delete", and "list" when possible.

Existing operation ids for this OpenAPI specification are as follows: ${existingIdsStr}.
The operation ID MUST be unique within this list of existing operation IDs. You can also use them as a reference to generate this operation id.`
}
