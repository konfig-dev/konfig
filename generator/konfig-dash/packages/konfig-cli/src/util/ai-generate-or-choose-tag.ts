import { Operation, TagObject, OperationObject } from 'konfig-lib'
import OpenAI from 'openai'

export async function aiGenerateOrChooseTag(
  operation: Operation,
  path: string,
  method: string,
  tags: TagObject[]
): Promise<string> {
  const openai = new OpenAI({ apiKey: process.env['OPENAI_API_KEY'] })
  const tagNames = tags.map((t) => t.name)

  // First, try and choose a tag programmatically without using OpenAI
  const tag = programaticallyChooseTag(operation, path, tagNames)
  if (tag !== undefined) return tag
  const systemPrompt = `You are a generator of tags for an OpenAPI specification. You generate tags in json format: "{ tag: ... }"`
  const prompt = generatePrompt(path, method, operation, tagNames)
  const response = (
    await openai.chat.completions.create({
      response_format: { type: 'json_object' },
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt },
      ],
      model: 'gpt-3.5-turbo-1106',
    })
  ).choices[0].message.content

  if (response === undefined || response === null || response === '')
    throw Error('OpenAI returned an empty response')

  const responseJson = JSON.parse(response)
  if (responseJson === undefined || responseJson === null)
    throw Error('OpenAI returned an invalid response: it is not a valid json')
  if (responseJson.tag === undefined || responseJson.tag === null)
    throw Error('OpenAI returned an invalid json: it does not contain a tag')

  tags.push({ name: responseJson.tag })
  return responseJson.tag
}

function programaticallyChooseTag(
  operation: Operation,
  path: string,
  tagNames: string[]
): string | undefined {
  for (const tag of tagNames) {
    if (path.toLowerCase().includes(tag.toLowerCase())) return tag
    if (
      operation.summary &&
      operation.summary.toLowerCase().includes(tag.toLowerCase())
    )
      return tag
  }
  return undefined
}

function generatePrompt(
  path: string,
  method: string,
  operation: Operation,
  tagNames: string[]
): string {
  const tagNamesStr = tagNames.join(', ').slice(0, -2)
  return `The following operation from my OpenAPI specification is missing a tag:
Path: ${path}
Method: ${method}
Summary: ${operation.summary}

Based on this information, choose the best tag from this list of existing tags on the OpenApi Specification: ${tagNamesStr}.
IF AND ONLY IF none of these existing tags are suitable, then generate a new tag. The new tag MUST be a single word, MUST be a noun, and MUST be singular. It MUST NOT be a combination of multiple words.

In general, a good tag name is related to the type of data being operated on (such as "Student"), or the type of operation being performed (such as "Authentication").

Respond with just the tag in json format: "{ tag: ... }" with no other information.`
}
