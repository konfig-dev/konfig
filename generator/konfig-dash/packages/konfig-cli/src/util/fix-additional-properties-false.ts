import { Spec, recurseObject } from 'konfig-lib'

/**
 * "additionalProperties: false" causes issue when mocking an API.  The
 * following type of errors will appear for no reason: "âœ–  error     Request
 * body must NOT have additional properties; found 'userId'" but userId is
 * actually a property in the schema so the error is just wrong.  To get around
 * this I added this rule to remove "additionalProperties: false" from the
 * schema.
 */
export function fixAdditionalPropertiesFalse({ spec }: { spec: Spec }): number {
  let numberOfAdditionalPropertiesFixed = 0
  recurseObject(spec.spec, ({ value: schema }) => {
    if (typeof schema !== 'object') return
    if (schema === null) return
    if (schema.type !== 'object') return
    if (schema.additionalProperties === undefined) return
    if (schema.additionalProperties === true) return
    delete schema.additionalProperties
    numberOfAdditionalPropertiesFixed++
  })
  return numberOfAdditionalPropertiesFixed
}
