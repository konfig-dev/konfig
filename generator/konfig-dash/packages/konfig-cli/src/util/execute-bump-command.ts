import { CliUx } from '@oclif/core'
import { getPackageVersion, setPackageVersion, KonfigYamlType } from 'konfig-lib'
import { generateKonfigYamlString } from '../util/generate-konfig-yaml-string'
import { incrementVersionWithPreservedTag } from '../util/increment-version-with-preserved-tag'
import semver from 'semver'
import * as fs from 'fs-extra'

export async function executeBumpCommand({
  filter,
  releaseType,
  loadedKonfigYaml,
  konfigYamlPath,
}: {
  filter: string[] | null
  releaseType: semver.ReleaseType
  loadedKonfigYaml: KonfigYamlType
  konfigYamlPath: string
}): Promise<void> {

  if (filter !== null)
    CliUx.ux.info(`Performing ${releaseType} bump of ${filter.join(', ')}`)
  else CliUx.ux.info(`Performing "${releaseType}" bump`)

  for (const [generatorName, generatorConfig] of [
    ...Object.entries(loadedKonfigYaml.generators),
    ...Object.entries(loadedKonfigYaml.additionalGenerators ?? []),
  ]) {
    if (filter !== null && !filter.includes(generatorName)) continue
    if ('disabled' in generatorConfig && generatorConfig.disabled) continue
    const currentVersion = getPackageVersion({ generatorConfig })
    const newVersion = incrementVersionWithPreservedTag(currentVersion, releaseType)
    CliUx.ux.info(`Bumping ${generatorName} SDK from ${currentVersion} --> ${newVersion}`)
    if (newVersion === null)
      CliUx.ux.error(`Could not compute new version for ${generatorName} SDK currently at ${currentVersion}`)
    setPackageVersion({ generatorConfig, version: newVersion })
  }
  CliUx.ux.action.start('Writing new version to konfig.yaml')
  fs.writeFileSync(konfigYamlPath, generateKonfigYamlString(loadedKonfigYaml))
  CliUx.ux.action.stop()
}
