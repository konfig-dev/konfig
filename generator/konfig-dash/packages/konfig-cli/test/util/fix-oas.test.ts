import { parseSpec, ResponsesObject, Spec } from 'konfig-lib'
import { fixOas } from '../../src/util/fix-oas'
import { Progress } from '../../src/util/fix-progress'

describe('fix-oas', () => {
  const responses: ResponsesObject = {
    '200': {
      description: 'OK',
      content: {
        'application/json': {
          schema: {
            type: 'string',
          },
        },
      },
    },
  }
  describe('custom modifications', () => {
    it('simple', async () => {
      const oas: Spec['spec'] = {
        openapi: '3.0.0',
        info: {
          title: 'Test API',
          description: 'Test',
          version: '1.0.0',
        },
        tags: [{ name: 'Tag' }],
        paths: {},
        components: {},
      }
      const spec = await parseSpec(JSON.stringify(oas))
      const progress = new Progress({
        progress: {},
        noSave: true,
      })
      await fixOas({
        spec,
        progress,
        konfigYaml: {
          defaultChangesetBumpType: 'patch',
          generators: {},
          specPath: '',
          fixConfig: { modify: { '$.info.description': 'New Test' } },
        },
        alwaysYes: true,
        auto: true,
      })
      expect(spec.spec).toMatchSnapshot()
    })
  })
  describe('disallowed-header-names', () => {
    it('request', async () => {
      const oas: Spec['spec'] = {
        openapi: '3.0.0',
        info: {
          title: 'Test API',
          description: 'Test',
          version: '1.0.0',
        },
        tags: [{ name: 'Tag' }],
        paths: {
          '/': {
            get: {
              tags: ['Tag'],
              operationId: 'Tag_get',
              parameters: [
                {
                  name: 'Content-Type',
                  in: 'header',
                },
                {
                  name: 'Accept',
                  in: 'header',
                },
                {
                  name: 'Authorization',
                  in: 'header',
                },
              ],
              responses,
            },
          },
        },
        components: {},
      }
      const spec = await parseSpec(JSON.stringify(oas))
      const progress = new Progress({
        progress: {
          securityParameters: {
            'Content-Type': { header: false },
            Accept: { header: false },
            Authorization: { header: false },
          },
        },
        noSave: true,
      })
      await fixOas({ spec, progress, alwaysYes: true, auto: true })
      expect(spec.spec).toMatchSnapshot()
    })

    it('response', async () => {
      const oas: Spec['spec'] = {
        openapi: '3.0.0',
        info: {
          title: 'Test API',
          description: 'Test',
          version: '1.0.0',
        },
        tags: [{ name: 'Tag' }],
        paths: {
          '/': {
            get: {
              tags: ['Tag'],
              operationId: 'Tag_get',
              parameters: [],
              responses: {
                '200': {
                  description: 'OK',
                  headers: {
                    'Content-Type': {},
                    Accept: {},
                    Authorization: {},
                  },
                },
              },
            },
          },
        },
        components: {},
      }
      const spec = await parseSpec(JSON.stringify(oas))
      const progress = new Progress({
        progress: {
          securityParameters: {
            'Content-Type': { header: false },
          },
        },
        noSave: true,
      })
      await fixOas({ spec, progress, alwaysYes: true, auto: true })
      expect(spec.spec).toMatchSnapshot()
    })
  })
  describe('oas3-valid-media-example', () => {
    it('fix-number-example', async () => {
      const oas: Spec['spec'] = {
        openapi: '3.0.0',
        info: {
          title: 'Test API',
          description: 'Test',
          version: '1.0.0',
        },
        tags: [{ name: 'Tag' }],
        paths: {
          '/': {
            get: {
              tags: ['Tag'],
              operationId: 'Tag_get',
              parameters: [
                {
                  name: 'lat',
                  in: 'query',
                  schema: { type: 'number' },
                  example: '-115.16988',
                },
              ],
              requestBody: {
                content: {
                  'application/json': {
                    schema: {
                      type: 'number',
                      example: '100',
                    },
                  },
                },
              },
              responses: {
                '200': {
                  description: 'OK',
                  headers: {
                    'Content-Type': {},
                    Accept: {},
                    Authorization: {},
                  },
                },
              },
            },
          },
        },
      }
      const spec = await parseSpec(JSON.stringify(oas))
      const progress = new Progress({
        progress: {
          securityParameters: {
            lat: {
              query: false,
            },
          },
        },
        noSave: true,
      })
      await fixOas({ spec, progress, alwaysYes: true, auto: true })
      expect(spec.spec).toMatchSnapshot()
    })
  })
  describe('oas3-schema', () => {
    it('fix mutual exclusiveness of example & examples field on media object', async () => {
      const oas: Spec['spec'] = {
        openapi: '3.0.0',
        info: {
          title: 'Test API',
          description: 'Test',
          version: '1.0.0',
        },
        tags: [{ name: 'Tag' }],
        paths: {
          '/': {
            get: {
              tags: ['Tag'],
              operationId: 'Tag_get',
              requestBody: {
                content: {
                  'application/json': {
                    schema: {
                      type: 'string',
                    },
                    example: '2',
                    examples: {
                      'example-0': { value: '1' },
                      'example-2': { value: '3' },
                    },
                  } as any,
                },
              },
              responses,
            },
          },
        },
        components: {},
      }
      const spec = await parseSpec(JSON.stringify(oas))
      const progress = new Progress({
        progress: {},
        noSave: true,
      })
      await fixOas({ spec, progress, alwaysYes: true, auto: true })
      expect(spec.spec).toMatchSnapshot()
    })
  })
  describe('ignore potential-incorrect-type', () => {
    const oas: Spec['spec'] = {
      openapi: '3.0.0',
      info: {
        title: 'Test API',
        description: 'Test',
        version: '1.0.0',
      },
      tags: [{ name: 'Tag' }],
      paths: {
        '/': {
          get: {
            operationId: 'Tag_get',
            requestBody: {
              content: {
                'application/json': {
                  schema: {
                    type: 'string',
                    example: '2',
                  },
                },
              },
            },
            responses,
          },
        },
      },
      components: {},
    }
    it('updates spec in place', async () => {
      const spec = await parseSpec(JSON.stringify(oas))
      const progress = new Progress({
        progress: { ignorePotentialIncorrectType: true },
        noSave: true,
      })
      await fixOas({ spec, progress, alwaysYes: true, auto: true })
      expect(spec.spec).toMatchSnapshot()
    })
  })
  describe('duplicate-tag-names', () => {
    const oas: Spec['spec'] = {
      openapi: '3.0.0',
      info: {
        title: 'Test API',
        description: 'Test',
        version: '1.0.0',
      },
      tags: [
        { name: 'Tag' },
        { name: 'tag' },
        { name: 'Tag with space' },
        { name: 'tag with space' },
      ],
      paths: {
        '/': {
          get: {
            tags: ['Tag'],
            operationId: 'Tag_get',
            responses,
          },
          options: {
            tags: ['Tag with space'],
            operationId: 'Tag_options',
            responses,
          },
          delete: {
            tags: ['tag with space'],
            operationId: 'Tag_delete',
            responses,
          },
          patch: {
            tags: ['tag with space'],
            operationId: 'Tag_head',
            responses,
          },
          head: {
            tags: ['tag'],
            operationId: 'Tag_head',
            responses,
          },
          post: {
            tags: ['Tag'],
            operationId: 'Tag_post',
            responses,
          },
        },
      },
    }
    it('deduplicate', async () => {
      const spec = await parseSpec(JSON.stringify(oas))
      const progress = new Progress({
        progress: {},
        noSave: true,
      })
      await fixOas({ spec, progress, alwaysYes: true, auto: true })
      expect(spec.spec).toMatchSnapshot()
    })
  })
  describe('object-with-no-properties', () => {
    const oas: Spec['spec'] = {
      openapi: '3.0.0',
      info: {
        title: 'Test API',
        description: 'Test',
        version: '1.0.0',
      },
      tags: [{ name: 'Tag' }],
      paths: {
        '/': {
          get: {
            operationId: 'Tag_get',
            requestBody: {
              content: {
                'application/json': {
                  schema: {
                    type: 'object',
                    example: {
                      hello: 'world',
                      onePlus: 2,
                    },
                  },
                },
              },
            },
            responses: {
              '200': {
                description: 'OK',
                content: {
                  'application/json': {
                    schema: {
                      type: 'object',
                      example: {
                        hello: 'world',
                        onePlus: 1,
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
      components: {},
    }
    it('generate schemas if not ignored', async () => {
      const spec = await parseSpec(JSON.stringify(oas))
      const progress = new Progress({
        progress: { ignoreObjectsWithNoProperties: false },
        noSave: true,
      })
      await fixOas({ spec, progress, alwaysYes: true, auto: true })
      expect(spec.spec).toMatchSnapshot()
    })
    it('generate schemas if ignored', async () => {
      const spec = await parseSpec(JSON.stringify(oas))
      const progress = new Progress({
        progress: { ignoreObjectsWithNoProperties: true },
        noSave: true,
      })
      await fixOas({ spec, progress, alwaysYes: true, auto: true })
      expect(spec.spec).toMatchSnapshot()
    })
  })
})
