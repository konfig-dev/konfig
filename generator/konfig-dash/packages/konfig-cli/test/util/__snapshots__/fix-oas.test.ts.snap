// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fix-oas array schema with missing "items" property items propety is added 1`] = `
{
  "components": {
    "schemas": {
      "MyArraySchema": {
        "items": {
          "description": "WARNING: Missing items property in array schema. Missing items property has been filled with this AnyType schema.",
        },
        "type": "array",
      },
      "TagGetRequest": {
        "properties": {
          "arrayProp": {
            "items": {
              "description": "WARNING: Missing items property in array schema. Missing items property has been filled with this AnyType schema.",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "TagGetResponse": {
        "type": "string",
      },
    },
  },
  "info": {
    "description": "Test",
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/": {
      "get": {
        "description": "Test",
        "operationId": "Tag_get",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/TagGetRequest",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TagGetResponse",
                },
              },
            },
            "description": "OK",
          },
        },
      },
    },
  },
  "tags": [
    {
      "name": "Tag",
    },
  ],
}
`;

exports[`fix-oas broken markdown links with whitespace between the [text] and (link) the whitespace is removed 1`] = `
{
  "components": {
    "schemas": {
      "TagGetResponse": {
        "type": "string",
      },
      "Test": {
        "description": "Test [link](https://example.com)",
      },
    },
  },
  "info": {
    "description": "Test [link](https://example.com)",
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/": {
      "get": {
        "description": "Test [link](https://example.com). Then [another link](https://example.com)",
        "operationId": "Tag_get",
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TagGetResponse",
                },
              },
            },
            "description": "OK",
          },
        },
      },
    },
  },
  "tags": [
    {
      "name": "Tag",
    },
  ],
}
`;

exports[`fix-oas custom modifications simple 1`] = `
{
  "components": {},
  "info": {
    "description": "New Test",
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {},
  "tags": [
    {
      "name": "Tag",
    },
  ],
}
`;

exports[`fix-oas disallowed-header-names request 1`] = `
{
  "components": {
    "schemas": {
      "TagGetResponse": {
        "type": "string",
      },
    },
  },
  "info": {
    "description": "Test",
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/": {
      "get": {
        "operationId": "Tag_get",
        "parameters": [
          {
            "in": "header",
            "name": "Authorization",
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TagGetResponse",
                },
              },
            },
            "description": "OK",
          },
        },
        "tags": [
          "Tag",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "Tag",
    },
  ],
}
`;

exports[`fix-oas disallowed-header-names response 1`] = `
{
  "components": {},
  "info": {
    "description": "Test",
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/": {
      "get": {
        "operationId": "Tag_get",
        "parameters": [],
        "responses": {
          "200": {
            "description": "OK",
            "headers": {
              "Authorization": {},
            },
          },
        },
        "tags": [
          "Tag",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "Tag",
    },
  ],
}
`;

exports[`fix-oas duplicate-tag-names deduplicate 1`] = `
{
  "components": {
    "schemas": {
      "TagDeleteResponse": {
        "type": "string",
      },
      "TagGetResponse": {
        "type": "string",
      },
      "TagHead200Response": {
        "type": "string",
      },
      "TagHeadResponse": {
        "type": "string",
      },
      "TagOptionsResponse": {
        "type": "string",
      },
      "TagPostResponse": {
        "type": "string",
      },
    },
  },
  "info": {
    "description": "Test",
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/": {
      "delete": {
        "operationId": "Tag_delete",
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TagDeleteResponse",
                },
              },
            },
            "description": "OK",
          },
        },
        "tags": [
          "tag with space",
        ],
      },
      "get": {
        "operationId": "Tag_get",
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TagGetResponse",
                },
              },
            },
            "description": "OK",
          },
        },
        "tags": [
          "Tag",
        ],
      },
      "head": {
        "operationId": "Tag_head",
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TagHeadResponse",
                },
              },
            },
            "description": "OK",
          },
        },
        "tags": [
          "Tag",
        ],
      },
      "options": {
        "operationId": "Tag_options",
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TagOptionsResponse",
                },
              },
            },
            "description": "OK",
          },
        },
        "tags": [
          "tag with space",
        ],
      },
      "patch": {
        "operationId": "Tag_head",
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TagHead200Response",
                },
              },
            },
            "description": "OK",
          },
        },
        "tags": [
          "tag with space",
        ],
      },
      "post": {
        "operationId": "Tag_post",
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TagPostResponse",
                },
              },
            },
            "description": "OK",
          },
        },
        "tags": [
          "Tag",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "Tag",
    },
    {
      "name": "tag with space",
    },
  ],
}
`;

exports[`fix-oas ignore potential-incorrect-type updates spec in place 1`] = `
{
  "components": {
    "schemas": {
      "TagGetRequest": {
        "example": "2",
        "type": "string",
      },
      "TagGetResponse": {
        "type": "string",
      },
    },
  },
  "info": {
    "description": "Test",
    "title": "Test API",
    "version": "1.0.0",
    "x-konfig-ignore": {
      "potential-incorrect-type": true,
    },
  },
  "openapi": "3.0.0",
  "paths": {
    "/": {
      "get": {
        "operationId": "Tag_get",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/TagGetRequest",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TagGetResponse",
                },
              },
            },
            "description": "OK",
          },
        },
      },
    },
  },
  "tags": [
    {
      "name": "Tag",
    },
  ],
}
`;

exports[`fix-oas oas3-schema fix mutual exclusiveness of example & examples field on media object 1`] = `
{
  "components": {
    "schemas": {
      "TagGetRequest": {
        "type": "string",
      },
      "TagGetResponse": {
        "type": "string",
      },
    },
  },
  "info": {
    "description": "Test",
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/": {
      "get": {
        "operationId": "Tag_get",
        "requestBody": {
          "content": {
            "application/json": {
              "examples": {
                "example-0": {
                  "value": "1",
                },
                "example-1": {
                  "value": "2",
                },
                "example-2": {
                  "value": "3",
                },
              },
              "schema": {
                "$ref": "#/components/schemas/TagGetRequest",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TagGetResponse",
                },
              },
            },
            "description": "OK",
          },
        },
        "tags": [
          "Tag",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "Tag",
    },
  ],
}
`;

exports[`fix-oas oas3-valid-media-example fix-number-example 1`] = `
{
  "components": {
    "schemas": {
      "TagGetRequest": {
        "example": 100,
        "type": "number",
      },
    },
  },
  "info": {
    "description": "Test",
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/": {
      "get": {
        "operationId": "Tag_get",
        "parameters": [
          {
            "example": -115.16988,
            "in": "query",
            "name": "lat",
            "schema": {
              "type": "number",
            },
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/TagGetRequest",
              },
            },
          },
        },
        "responses": {
          "200": {
            "description": "OK",
            "headers": {
              "Authorization": {},
            },
          },
        },
        "tags": [
          "Tag",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "Tag",
    },
  ],
}
`;

exports[`fix-oas object-with-no-properties generate schemas if ignored 1`] = `
{
  "components": {
    "schemas": {
      "TagGetRequest": {
        "example": {
          "hello": "world",
          "onePlus": 2,
        },
        "properties": {
          "hello": {
            "example": "world",
            "type": "string",
          },
          "onePlus": {
            "example": 2,
            "type": "number",
          },
        },
        "type": "object",
      },
      "TagGetResponse": {
        "example": {
          "hello": "world",
          "onePlus": 1,
        },
        "type": "object",
      },
    },
  },
  "info": {
    "description": "Test",
    "title": "Test API",
    "version": "1.0.0",
    "x-konfig-ignore": {
      "object-with-no-properties": true,
    },
  },
  "openapi": "3.0.0",
  "paths": {
    "/": {
      "get": {
        "operationId": "Tag_get",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/TagGetRequest",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TagGetResponse",
                },
              },
            },
            "description": "OK",
          },
        },
      },
    },
  },
  "tags": [
    {
      "name": "Tag",
    },
  ],
}
`;

exports[`fix-oas object-with-no-properties generate schemas if not ignored 1`] = `
{
  "components": {
    "schemas": {
      "TagGetRequest": {
        "example": {
          "hello": "world",
          "onePlus": 2,
        },
        "properties": {
          "hello": {
            "example": "world",
            "type": "string",
          },
          "onePlus": {
            "example": 2,
            "type": "number",
          },
        },
        "type": "object",
      },
      "TagGetResponse": {
        "example": {
          "hello": "world",
          "onePlus": 1,
        },
        "properties": {
          "hello": {
            "example": "world",
            "type": "string",
          },
          "onePlus": {
            "example": 1,
            "type": "number",
          },
        },
        "type": "object",
      },
    },
  },
  "info": {
    "description": "Test",
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/": {
      "get": {
        "operationId": "Tag_get",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/TagGetRequest",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TagGetResponse",
                },
              },
            },
            "description": "OK",
          },
        },
      },
    },
  },
  "tags": [
    {
      "name": "Tag",
    },
  ],
}
`;

exports[`fix-oas request media type objects with missing "schema" property are fixed 1`] = `
{
  "components": {
    "requestBodies": {
      "TagRequestBody": {
        "content": {
          "application/json": {
            "example": "Should have a schema",
            "schema": {
              "description": "WARNING: Missing schema in media type object. Missing schema has been filled with this AnyType schema.",
            },
          },
        },
      },
    },
    "schemas": {
      "TagGetResponse": {
        "type": "string",
      },
    },
  },
  "info": {
    "description": "Test",
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/": {
      "get": {
        "description": "Test",
        "operationId": "Tag_get",
        "requestBody": {
          "$ref": "#/components/requestBodies/TagRequestBody",
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TagGetResponse",
                },
              },
            },
            "description": "OK",
          },
        },
      },
    },
  },
  "tags": [
    {
      "name": "Tag",
    },
  ],
}
`;

exports[`fix-oas security.in fields with uppercase characters are lowercased 1`] = `
{
  "components": {
    "schemas": {
      "TagGetResponse": {
        "type": "string",
      },
    },
    "securitySchemes": {
      "openapi_authorization": {
        "in": "header",
        "name": "Authorization",
        "type": "apiKey",
      },
    },
  },
  "info": {
    "description": "Test",
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/": {
      "get": {
        "description": "Test",
        "operationId": "Tag_get",
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TagGetResponse",
                },
              },
            },
            "description": "OK",
          },
        },
      },
    },
  },
  "tags": [
    {
      "name": "Tag",
    },
  ],
}
`;
